chef/cookbooks/utils/libraries/service.rb

```diff
@@ -0,0 +1,73 @@
+#
+# Copyright 2017, SUSE Linux GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# As we want to override the restart method to either call it or skip it,
+# we just monkey patch the Chef::Provider::Service to override the action_restart
+# which is the common method for all the children Service classes
+class Chef
+  class Provider
+    class Service < Chef::Provider
+      def action_restart
+        # how do we know if its allowed to restart? We use the cookbook name as the key
+
+        # as both keystone and horizon use apache2, they trigger a restart of it when something
+        # changes instead of having their own services
+        # unless we change all of their restart triggers to be a reload, this is the best we can do
+        # to avoid restarts :(
+        if cookbook_name == "apache2"
+          disallow_restart = [
+            node.fetch("horizon".to_s, {}).fetch(:disallow_restart, false),
+            node.fetch("keystone", {}).fetch(:disallow_restart, false)
+          ].any?
+        else
+          disallow_restart = node.fetch(cookbook_name.to_s, {})[:disallow_restart] || false
+        end
+        # safeguards
+        node.set[cookbook_name] = {} unless node[cookbook_name]
+        node.set[cookbook_name][:requires_restart] = {} \
+          unless node.set[cookbook_name][:requires_restart]
+
+        if disallow_restart
+          Chef::Log.info("Disallowing restart for #{@new_resource.name} due to flag")
+          # Store more data about the service attempted restart
+          # use the service as the key name for the data for easy removal
+          # force it to be a simple hash instead of a node attribute if you want the
+          # update method to actually update. If you update the attribute directly it doesnt
+          # update but overwrites! The joys of chef!
+          requires_restart_hash = node[cookbook_name][:requires_restart].to_h
+          node.set[cookbook_name][:requires_restart] = requires_restart_hash.update(
+            @new_resource.name.to_sym => {
+              pacemaker_service: false,
+              timestamp: Time.now.getutc
+            }
+          )
+        else
+          # from this point this is a copy of the default method
+          converge_by("restart service #{@new_resource}") do
```

Comment:
from `Chef::Provider::Service#action_restart` https://github.com/SUSE-Cloud/chef/blob/10-stable-suse/chef/lib/chef/provider/service.rb#L113

Reply:
I tired to but it seems impossible to me. I may not know enough ruby as to know how to do it, but I expected a super to work here, but that doesnt work, it claims the parent class has no disallow_restart method or something like that.

As the Chef::Provider::Service then gets overload with the specific Service class (InitD, SystemD, etc..) there maybe something special to do here that Im missing. We should get a ruby expert to hlep here methinks.

Reply:
I agree with @matelakat that we should avoid copy'n'paste if we possibly can.  I don't think you can use inheritance because Chef expects to use this exact provider class.  However can't you use `Module#prepend` as described in that StackOverflow answer?  It looks like a clean and simple solution.

Reply:
@aspiers it turned out not to be a copy-paste actually, so the comment was modified to say "this is a modified version of the original method" also asked @rsalevsky on that prepend technique and the decision was to keep it as it is for now.